### PR TITLE
FSET-1649 Notify candidate of invitation to FSAC event

### PR DESCRIPTION
**Pull request by vglushak**

### CODE REVIEW

File: app/connectors/EmailClient.scala [hunk 1]
```diff
@@ -209,6 +209,18 @@ trait EmailClient extends WSHttp {
   def sendApplicationExtendedToSdip(to: String, name: String)(implicit hc: HeaderCarrier): Future[Unit] =
     sendEmail(to, "fset_faststream_app_converted_to_sdip_confirmation", Map("name" -> name))
 
+  def sendCandidateInvitationRequestToFSAC(to: String, name: String,
```
Comment (mgilgar) [inline app/connectors/EmailClient.scala]: Shouldn't this be sendCandidateConfirmationRequestToFSAC?

File: app/model/stc/EmailEvents.scala [hunk 2]
```diff
@@ -32,4 +32,9 @@ object EmailEvents {
   case class AdjustmentsConfirmed(to: String, name: String, etrayAdjustments: String, videoAdjustments: String) extends EmailEvent
   case class AdjustmentsChanged(to: String, name: String, etrayAdjustments: String, videoAdjustments: String) extends EmailEvent
   case class ApplicationConvertedToSdip(to: String, name: String) extends EmailEvent
+
+  case class CandidateAllocationConfirmed(to: String, name: String, eventDate: String, eventTime: String) extends EmailEvent
+  case class CandidateAllocationRequestConfirmation(to: String, name: String, eventDate: String, eventTime: String,
```
Comment (mgilgar) [inline app/model/stc/EmailEvents.scala]: I think should be verbs in the past. Maybe CandidateAllocationConfirmationRequested.

File: app/services/allocation/AssessorAllocationService.scala [hunk 3]
```diff
@@ -16,31 +16,50 @@
...
+import uk.gov.hmrc.play.http.HeaderCarrier
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
 object AssessorAllocationService extends AssessorAllocationService {
-  def allocationRepo = repositories.assessorAllocationRepository
-  override val eventsService = EventsService
-  def candidateAllocationRepo = repositories.candidateAllocationRepository
+  val allocationRepo: AssessorAllocationMongoRepository = repositories.assessorAllocationRepository
```
Comment (mgilgar) [inline app/services/allocation/AssessorAllocationService.scala]: To distinguish clearly from candidateAllocationRepo, I would call it assessorAllocationRepo

File: app/services/allocation/AssessorAllocationService.scala [hunk 4]
```diff
@@ -16,31 +16,50 @@
...
+  val candidateAllocationRepo: CandidateAllocationMongoRepository = repositories.candidateAllocationRepository
+  val applicationRepo: GeneralApplicationRepository = repositories.applicationRepository
+  val eventsService = EventsService
```
Comment (mgilgar) [inline app/services/allocation/AssessorAllocationService.scala]: I would put all services together and all clients together.

File: app/services/allocation/AssessorAllocationService.scala [hunk 5]
```diff
@@ -56,27 +75,75 @@ trait AssessorAllocationService {
...
-  def allocateCandidates(newAllocations: command.CandidateAllocations): Future[Unit] = {
-    getCandidateAllocations(newAllocations.eventId).flatMap { existingAllocation =>
-      existingAllocation.allocations match {
-        case Nil => candidateAllocationRepo.save(persisted.CandidateAllocation.fromCommand(newAllocations)).map(_ => ())
-        case _ => updateExistingAllocations(existingAllocation, newAllocations).map(_ => ())
+  private val dateFormat = "dd MMMM YYYY"
+  private val timeFormat = "HH:mma"
+
+
+  def allocateCandidates(newAllocations: command.CandidateAllocations)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = {
```
Comment (mgilgar) [inline app/services/allocation/AssessorAllocationService.scala]: Why do you need the allocateCandidates method in AssessorAllocationService.scala file? I think we need a separate CandidateAllocationService in the same way we have a candidateAllocationRepo and assessorAllocationRepo.

File: test/services/allocation/AssessorAllocationServiceSpec.scala [hunk 6]
```diff
@@ -80,14 +84,35 @@ class AssessorAllocationServiceSpec extends BaseServiceSpec {
     }
   }
 
+  "Allocate candidate" must {
```
Comment (mgilgar) [inline test/services/allocation/AssessorAllocationServiceSpec.scala]: If you decide to move allocate candidates to its own service, you will have to create a separate spec I guess for this test.

File: app/connectors/EmailClient.scala (hunk 1 above)
Comment (vglushak) [inline app/connectors/EmailClient.scala]: yes.

File: app/model/stc/EmailEvents.scala (hunk 2 above)
Comment (vglushak) [inline app/model/stc/EmailEvents.scala]: yes.

File: app/services/allocation/AssessorAllocationService.scala (hunk 3 above)
Comment (vglushak) [inline app/services/allocation/AssessorAllocationService.scala]: ok.

File: app/services/allocation/AssessorAllocationService.scala (hunk 5 above)
Comment (vglushak) [inline app/services/allocation/AssessorAllocationService.scala]: Well, yes, but:
- out of scope of this task.
- taking into account current test coverage I would prefer not to do unrelated refactorings.

Comment (mgilgar) [inline app/services/allocation/AssessorAllocationService.scala]: It seems this is new code: in this case it is not refactoring. Or do you move it inside the file?

Comment (vglushak) [inline app/services/allocation/AssessorAllocationService.scala]: I've added implicit arguments set, but method already exists.

Comment (kjharkin) [inline app/services/allocation/AssessorAllocationService.scala]: This should be moved but we can do so when we revisit the allocations.

File: app/services/allocation/AssessorAllocationService.scala [hunk 7]
```diff
@@ -16,67 +16,133 @@
...
+      val eventTime = event.startTime.toString(timeFormat)
+      val deadlineDateTime = event.date.minusDays(10).toString(dateFormat)
+
+
+      getCandidateAllocations(newAllocations.eventId).flatMap { existingAllocation =>
+        existingAllocation.allocations match {
+          case Nil =>
+            Future.sequence(newAllocations.allocations.map(sendCandidateEmail(_, eventDate, eventTime, deadlineDateTime)))
+              .flatMap { _ =>
+                candidateAllocationRepo.save(persisted.CandidateAllocation.fromCommand(newAllocations)).map(_ => ())
```
Comment (kjharkin) [inline app/services/allocation/AssessorAllocationService.scala]: I think the order here is wrong - we should save the allocations before sending the email, otherwise if things go wrong we may invite a candidate without having an allocation for them - could be embarrassing.

File: app/services/allocation/AssessorAllocationService.scala [hunk 8]
```diff
@@ -16,67 +16,133 @@
...
+              }
+          case _ =>
+            val existingIds = existingAllocation.allocations.map(_.id)
+            Future.sequence(
+              newAllocations.allocations
+                .filter(alloc => !existingIds.contains(alloc.id))
+                .map(sendCandidateEmail(_, eventDate, eventTime, deadlineDateTime))
+            ).flatMap { _ =>
+              updateExistingAllocations(existingAllocation, newAllocations).map(_ => ())
```
Comment (kjharkin) [inline app/services/allocation/AssessorAllocationService.scala]: Same here

File: app/services/allocation/AssessorAllocationService.scala [hunk 9]
```diff
@@ -16,67 +16,133 @@
...
+          val res = authProviderClient.findByUserIds(Seq(candidate.userId)).map { candidates =>
+            candidates.map { candidate =>
+              candidateAllocation.status match {
+                case AllocationStatuses.UNCONFIRMED =>
+                  CandidateAllocationConfirmationRequest(candidate.email, candidate.name, eventDate, eventTime, deadlineDateTime)
+                case AllocationStatuses.CONFIRMED =>
+                  CandidateAllocationConfirmed(candidate.email, candidate.name, eventDate, eventTime)
+              }
+            }
+          } recover { case _ => throw new RuntimeException("Was not able to retrieve user details.") }
```
Comment (kjharkin) [inline app/services/allocation/AssessorAllocationService.scala]: whose user details?

File: app/services/allocation/AssessorAllocationService.scala [hunk 10]
```diff
@@ -16,67 +16,133 @@
...
+          res.asInstanceOf[Future[StcEvents]]
+        }
+      case None => throw new RuntimeException("Can not find user application")
```
Comment (kjharkin) [inline app/services/allocation/AssessorAllocationService.scala]: which user application?
